core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -573,9 +576,14 @@ class UnifiedLog(@volatile var logStartOffset: Long,
       case None =>
         if (keepPartitionMetadataFile) {
           _topicId = Some(topicId)
-          if (!partitionMetadataFile.exists()) {
-            partitionMetadataFile.record(topicId)
-            scheduler.schedule("flush-metadata-file", maybeFlushMetadataFile)
+          partitionMetadataFile match {
+            case Some(partitionMetadataFile) =>
+              if (!partitionMetadataFile.exists()) {
+                partitionMetadataFile.record(topicId)
+                scheduler.schedule("flush-metadata-file", maybeFlushMetadataFile)
+              }
+            case _ => warn(s"The topic id $topicId will not be persisted to the partition metadata file " +
+              "since the field is not initialized")
```

Comment:
since the field is not initialized => since the partition is deleted?

Reply:
That looks good to me, I'll revise this log and update the PR after resolving the other comments.